packages/iov-ethereum/src/ethereumconnection.ts

```diff
@@ -774,36 +802,160 @@ export class EthereumConnection implements AtomicSwapConnection {
       const timeoutEnd = timeoutBegin + 32;
       const amountBegin = timeoutEnd;
       const amountEnd = amountBegin + 32;
+      const preimageBegin = amountEnd;
+      const preimageEnd = preimageBegin + 32;
+      const stateBegin = preimageEnd;
+      const stateEnd = stateBegin + 32;
 
       const resultArray = Encoding.fromHex(normalizeHex(swapsResponse.result));
-      const sender = toChecksummedAddress(Abi.decodeAddress(resultArray.slice(senderBegin, senderEnd)));
-      const recipient = toChecksummedAddress(
-        Abi.decodeAddress(resultArray.slice(recipientBegin, recipientEnd)),
-      );
-      const hash = resultArray.slice(hashBegin, hashEnd) as Hash;
-      const timeoutHeight = new BN(resultArray.slice(timeoutBegin, timeoutEnd)).toNumber();
-      const amount = {
-        quantity: new BN(resultArray.slice(amountBegin, amountEnd)).toString(),
-        fractionalDigits: constants.primaryTokenFractionalDigits,
-        tokenTicker: constants.primaryTokenTicker,
+      const swapData = {
+        id: query.swapid,
+        sender: toChecksummedAddress(Abi.decodeAddress(resultArray.slice(senderBegin, senderEnd))),
+        recipient: toChecksummedAddress(Abi.decodeAddress(resultArray.slice(recipientBegin, recipientEnd))),
+        hash: resultArray.slice(hashBegin, hashEnd) as Hash,
+        amounts: [
+          {
+            quantity: new BN(resultArray.slice(amountBegin, amountEnd)).toString(),
+            fractionalDigits: constants.primaryTokenFractionalDigits,
+            tokenTicker: constants.primaryTokenTicker,
+          },
+        ] as ReadonlyArray<Amount>,
+        timeout: {
+          height: new BN(resultArray.slice(timeoutBegin, timeoutEnd)).toNumber(),
+        },
       };
 
-      return [
+      const kind = Abi.decodeSwapProcessState(resultArray.slice(stateBegin, stateEnd));
+      let swap: AtomicSwap;
+      if (kind === SwapProcessState.Open) {
+        swap = ({
+          kind: kind,
+          data: swapData,
+        } as any) as OpenSwap;
```

Comment:
why are those cast necessary?

Reply:
Yeah, this is the bit I'm most unhappy with. I haven't got to the bottom of why yet, and this `as any as X` stuff is definitely a hack that should not stay in.